src/Explorer/Web/ContractListView.hs

```diff
@@ -6,67 +6,194 @@ module Explorer.Web.ContractListView
+renderCLVRs :: ContractListView -> Html
+
+renderCLVRs (ContractListView timeNow retrievalTime mbPage clvrs) = baseDoc "Marlowe Contract List" $ do
+  let
+    page = normalizePage (length clvrs) mbPage
+    pageOfClvrs = takePage page clvrs
+  renderTime timeNow retrievalTime
+  table $ do
     tr $ do
       th $ b "Contract ID"
+      th $ b "Role token minting policy"
       th $ b "Block No"
       th $ b "Slot No"
+      th $ b ""
     let makeRow clvr = do
           let cid = clvrContractId clvr
           tr $ do
             td $ a ! href (toValue $ generateLink "contractView" [("tab", "info"), ("contractId", cid)])
               $ string cid
+            td $ renderStr . clvrRoleMintingPolicyId $ clvr
             td $ toHtml . clvrBlock $ clvr
             td $ toHtml . clvrSlot $ clvr
-    forM_ clvrs makeRow
+            td $ a ! href (toValue $ clvrBlockExplLink clvr) $ string "Explore"
+    forM_ pageOfClvrs makeRow
+  renderNavBar page (length clvrs)
+
+renderCLVRs (ContractListViewError msg) =
+  baseDoc "An error occurred" $ string ("Error: " <> msg)
+
+
+generateLink' :: Int -> String -> Html
+generateLink' targetPage label =
+  a ! href (toValue $ generateLink "listContracts" [("page", show targetPage)]) $ string label
+
+space :: Html
+space = preEscapedToHtml ("&nbsp;&nbsp;" :: String)
+
+calcLastPage :: Int -> Int
+calcLastPage numContracts = div numContracts pageLength + 1
+
+renderNavBar :: Int -> Int -> Html
+renderNavBar page numContracts = p $ do
+  generateFirst page 1 "<<"
+  space
+  generatePrev page (page - 1) "<"
+  space
+  renderNavNumbers page numContracts
+  generateNext page (page + 1) ">"
+  space
+  generateLast page lastPage ">>"
+  space
+  string $ printf "| %d-%d contracts shown out of %d, (page %d out of %d)"
+    ((page - 1) * pageLength + 1)
+    (min (((page - 1) * pageLength) + pageLength) numContracts)
+    numContracts
+    page lastPage
+
+  where
+    lastPage = calcLastPage numContracts
+
+    generateFirst curPage targetPage label
+      | curPage == 1 = return ()
+      | otherwise = generateLink' targetPage label
+    generatePrev curPage targetPage label
+      | curPage == 1 = return ()
+      | otherwise = generateLink' targetPage label
+    generateNext curPage targetPage label
+      | curPage == lastPage = return ()
+      | otherwise = generateLink' targetPage label
+    generateLast curPage targetPage label
+      | curPage == lastPage = return ()
+      | otherwise = generateLink' targetPage label
+
+
+renderNavNumbers :: Int -> Int -> Html
+renderNavNumbers page numContracts = mapM_ generateNumLink generatePageList
+  where
+    lastPage = calcLastPage numContracts
+
+    generatePageList
+      | page < 6 = [1..11]
+      | page > lastPage - 5 = [lastPage - 10..lastPage]
+      | otherwise = [page - 5..page + 5]
+
+    generateNumLink curLink
+      | page == curLink = (string . show $ curLink) >> space
+      | otherwise = generateLink' curLink (show curLink) >> space
```

Comment:
Not the same anymore. `generateNavLink` takes 3 arguments now because it's not the same comparison as this code. Having trouble writing one function to do both that looks nice because we need the last arg twice when it's used by `renderNavNumbers`.
I guess I could do this: `(\p -> generateNavLink comparisonPage someLabel p p)` for the `mapM_` but it's not pretty.